File: packages/numfmt/src/parse.ts
/**
 * A format code, taken apart.
 *
 * `#,##0.00;[Red](#,##0.00);"—";@` is four formats in one string: what a
 * positive number looks like, a negative, a zero, and text. Excel's little
 * language has been added to for thirty years and reads like it — the same
 * letter means different things in different places, and the only way to know
 * which is to look at what is around it.
 *
 * Two of those ambiguities matter enough to name:
 *
 * - **`m` is a month or a minute.** After an hour or before a second it is
 *   minutes; everywhere else it is months. `h:m:s` is a clock and `m/d/yyyy`
 *   is a date, and the letters are the same.
 * - **`/` is a fraction or a date separator.** In `# ?/?` it divides; in
 *   `d/m/yyyy` it is a stroke between numbers.
 *
 * Both are decided by what kind of section it is, which is decided first.
 */

export type Token =
  | { kind: 'digit'; placeholder: '0' | '#' | '?' }
  | { kind: 'decimal' }
  /** A comma between digits, which groups thousands. */
  | { kind: 'group' }
  /** A comma after the digits, which divides by a thousand for each one. */
  | { kind: 'scale'; by: number }
  | { kind: 'percent' }
  | { kind: 'literal'; text: string }
  /** `*x` — repeat the character until the cell is full, which only a grid knows. */
  | { kind: 'fill'; char: string }
  /** `_x` — a space as wide as the character, for lining columns up. */
  | { kind: 'pad'; char: string }
  /** `@` — where the text goes in a text section. */
  | { kind: 'text' }
  /** `E+` or `e-`: the letter is kept because Excel echoes the case it was given. */
  | { kind: 'exponent'; sign: '+' | '-'; letter: 'e' | 'E' }
  | { kind: 'fraction' }
  | { kind: 'date'; code: string }
  /** `[h]`, `[mm]`, `[ss]` — a unit that counts past its own wrap. */
  | { kind: 'elapsed'; code: string }

export interface Condition {
  operator: '<' | '<=' | '>' | '>=' | '=' | '<>'
  value: number
}

export interface Section {
  tokens: Token[]
  /** `[Red]` and the rest, as written; the caller decides what red is. */
  color: string | null
  /** `[>=100]` — which numbers this section is for. */
  condition: Condition | null
  kind: 'number' | 'date' | 'text'
}

export interface NumberFormat {
  sections: Section[]
  /** The code it was read from, for a writer that keeps what it was given. */
  code: string
}

const COLORS = new Set(['black', 'blue', 'cyan', 'green', 'magenta', 'red', 'white', 'yellow'])

/** Splits on the semicolons that separate sections, ignoring the quoted ones. */
function sectionsOf(code: string): string[] {
  const parts: string[] = []
  let current = ''
  let quoted = false
  let bracketed = false

  for (let at = 0; at < code.length; at += 1) {
    const char = code[at] ?? ''

    if (char === '\\' && !quoted) {
      current += char + (code[at + 1] ?? '')
      at += 1
      continue
    }
    if (char === '"') quoted = !quoted
    if (!quoted && char === '[') bracketed = true
    if (!quoted && char === ']') bracketed = false

    if (char === ';' && !quoted && !bracketed) {
      parts.push(current)
      current = ''
      continue
    }

    current += char
  }

  parts.push(current)
  return parts
}

const DATE_LETTERS = /[ymdhs]/iu

/**
 * Whether a section formats a date.
 *
 * Decided before anything else is read, because it is what makes `m` a month
 * and `/` a stroke. Letters inside quotes or brackets do not count: `0 "months"`
 * is a number, however much it talks about time.
 */
function looksLikeDate(body: string): boolean {
  const bare = body
    .replace(/"[^"]*"/gu, '')
    .replace(/\\./gu, '')
    .replace(/\[(?!h+\]|m+\]|s+\])[^\]]*\]/giu, '')

  return DATE_LETTERS.test(bare) || /\[(?:h+|m+|s+)\]/iu.test(bare)
}

/**
 * A replacement that leaves quoted runs alone.
 *
 * Everything in this language can appear inside quotes and mean itself, so a
 * scan that ignores them reads `"[" @ "]"` as a bracketed annotation and
 * deletes the text format.
 */
function replaceOutsideQuotes(
  body: string,
  pattern: RegExp,
  replace: (whole: string, inside: string) => string,
): string {
  const parts = body.split(/("[^"]*")/u)

  return parts
    .map((part) =>
      part.startsWith('"')
        ? part
        : part.replace(pattern, (whole, inside: string) => replace(whole, inside)),
    )
    .join('')
}

/** The condition, colour and currency a section states in brackets. */
function readBrackets(body: string): {
  rest: string
  color: string | null
  condition: Condition | null
  currency: string | null
} {
  let color: string | null = null
  let condition: Condition | null = null
  let currency: string | null = null

  // Quoted runs are stepped over: `"[" @ "]"` is a text format with square
  // brackets in it, not a colour and a condition.
  const rest = replaceOutsideQuotes(body, /\[([^\]]*)\]/gu, (whole, inside) => {
    const lower = inside.toLowerCase()

    // An elapsed unit is a token rather than an annotation, so it stays.
    if (/^[hms]+$/iu.test(inside)) return whole

    if (COLORS.has(lower) || /^color\s*\d+$/u.test(lower)) {
      color = inside
      return ''
    }

    const test = /^([<>=]=?|<>)\s*(-?[\d.]+)$/u.exec(inside)
    if (test !== null) {
      const operator = test[1] as Condition['operator']
      condition = { operator, value: Number(test[2]) }
      return ''
    }

    // `[$€-407]` states a currency and a locale; `[$-409]` states only the
    // locale, which decides month names in Excel and is ignored here. The
    // difference is whether there is a symbol before the dash.
    const money = /^\$([^-]*)(?:-.*)?$/u.exec(inside)
    if (money !== null) {
      const symbol = money[1] ?? ''
      currency = symbol === '' ? null : symbol
      return ''
    }

    // Anything else in brackets — a locale on its own, a calendar — says
    // nothing this can act on and nothing worth showing.
    return ''
  })

  return { rest, color, condition, currency }
}

/** Runs of the same date letter are one token: `yyyy` is a year, not four. */
function dateRun(body: string, at: number): string {
  const letter = body[at] ?? ''
  let end = at

  while ((body[end + 1] ?? '').toLowerCase() === letter.toLowerCase()) end += 1
  return body.slice(at, end + 1)
}

function tokenise(body: string, kind: Section['kind'], currency: string | null): Token[] {
  const tokens: Token[] = []
  const push = (token: Token) => tokens.push(token)

  for (let at = 0; at < body.length; at += 1) {
    const char = body[at] ?? ''

    if (char === '"') {
      const end = body.indexOf('"', at + 1)
      push({ kind: 'literal', text: body.slice(at + 1, end === -1 ? body.length : end) })
      at = end === -1 ? body.length : end
      continue
    }

    if (char === '\\') {
      push({ kind: 'literal', text: body[at + 1] ?? '' })
      at += 1
      continue
    }

    if (char === '_') {
      push({ kind: 'pad', char: body[at + 1] ?? ' ' })
      at += 1
      continue
    }

    if (char === '*') {
      push({ kind: 'fill', char: body[at + 1] ?? ' ' })
      at += 1
      continue
    }

    if (char === '[') {
      const end = body.indexOf(']', at)
      const inside = body.slice(at + 1, end === -1 ? body.length : end)
      if (/^[hms]+$/iu.test(inside)) push({ kind: 'elapsed', code: inside })
      at = end === -1 ? body.length : end
      continue
    }

    if (char === '0' || char === '#' || char === '?') {
      push({ kind: 'digit', placeholder: char })
      continue
    }

    if (char === '.') {
      // A date's dot separates — `dd.mm.yy` — unless it is the point before a
      // fraction of a second, which is decided by what follows it.
      const fractionOfSecond = kind === 'date' && /^\.0+/u.test(body.slice(at))
      push(
        kind === 'date' && !fractionOfSecond ? { kind: 'literal', text: '.' } : { kind: 'decimal' },
      )
      continue
    }

    if (char === '%') {
      push({ kind: 'percent' })
      continue
    }

    if (char === '@') {
      push({ kind: 'text' })
      continue
    }

    if ((char === 'E' || char === 'e') && (body[at + 1] === '+' || body[at + 1] === '-')) {
      push({ kind: 'exponent', sign: body[at + 1] === '+' ? '+' : '-', letter: char })
      at += 1
      continue
    }

    if (char === ',') {
      // In a date a comma is punctuation — `mmm d, yyyy`. In a number it is
      // either grouping or scaling, and which one depends on what follows, so
      // the scaling ones are folded together afterwards.
      push(kind === 'date' ? { kind: 'literal', text: ',' } : { kind: 'group' })
      continue
    }

    if (char === '/' && kind === 'number') {
      push({ kind: 'fraction' })
      continue
    }

    if (kind === 'date' && DATE_LETTERS.test(char)) {
      const run = dateRun(body, at)
      push({ kind: 'date', code: run })
      at += run.length - 1
      continue
    }

    if (kind === 'date' && /^(?:AM\/PM|A\/P)/iu.test(body.slice(at))) {
      const run = body.slice(at).toUpperCase().startsWith('AM/PM')
        ? body.slice(at, at + 5)
        : body.slice(at, at + 3)
      push({ kind: 'date', code: run })
      at += run.length - 1
      continue
    }

    push({ kind: 'literal', text: char })
  }

  if (currency !== null) tokens.unshift({ kind: 'literal', text: currency })
  return foldScaling(tokens)
}

/**
 * Commas that come after every digit, turned into division.
 *
 * `#,##0,` shows thousands and `#,##0,,` shows millions. The trailing ones are
 * the scaling; the ones between digits are grouping, and telling them apart is
 * a matter of what comes after.
 */
function foldScaling(tokens: readonly Token[]): Token[] {
  const folded: Token[] = []

  for (const [index, token] of tokens.entries()) {
    if (token.kind !== 'group') {
      folded.push(token)
      continue
    }

    const later = tokens.slice(index + 1)
    const trailing = !later.some((one) => one.kind === 'digit' || one.kind === 'decimal')

    if (!trailing) {
      folded.push(token)
      continue
    }

    const last = folded[folded.length - 1]
    if (last?.kind === 'scale') last.by *= 1000
    else folded.push({ kind: 'scale', by: 1000 })
  }

  return folded
}

/**
 * The AM/PM marker turns a 24-hour format into a 12-hour one.
 *
 * Stated here rather than found later because it changes what `h` means, and
 * the hour is written long before the marker is reached.
 */
export const isTwelveHour = (section: Section): boolean =>
  section.tokens.some((token) => token.kind === 'date' && /^(?:AM\/PM|A\/P)$/iu.test(token.code))

export function parseFormat(code: string): NumberFormat {
  const sections = sectionsOf(code).map((body) => {
    const { rest, color, condition, currency } = readBrackets(body)
    const kind: Section['kind'] = looksLikeDate(rest)
      ? 'date'
      : rest.includes('@')
        ? 'text'
        : 'number'

    return { tokens: tokenise(rest, kind, currency), color, condition, kind }
  })

  return { sections, code }
}

File: crates/formula/src/numfmt/parse.rs
//! A format code, taken apart.
//!
//! `#,##0.00;[Red](#,##0.00);"—";@` is four formats in one string: what a
//! positive number looks like, a negative, a zero, and text. Excel's little
//! language has been added to for thirty years and reads like it — the same
//! letter means different things in different places, and the only way to
//! know which is to look at what is around it.
//!
//! Two of those ambiguities matter enough to name:
//!
//! - **`m` is a month or a minute.** After an hour or before a second it is
//!   minutes; everywhere else it is months. `h:m:s` is a clock and `m/d/yyyy`
//!   is a date, and the letters are the same.
//! - **`/` is a fraction or a date separator.** In `# ?/?` it divides; in
//!   `d/m/yyyy` it is a stroke between numbers.
//!
//! Both are decided by what kind of section it is, which is decided first.

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Digit(char),
    Decimal,
    /// A comma between digits, which groups thousands.
    Group,
    /// A comma after the digits, dividing by a thousand for each one.
    Scale(f64),
    Percent,
    Literal(String),
    /// `*x` — repeat until the cell is full, which only a grid can do.
    Fill(char),
    /// `_x` — a space as wide as the character, for lining columns up.
    Pad(char),
    /// `@` — where the text goes in a text section.
    Text,
    /// `E+` or `e-`: which of the two, and the letter, because Excel echoes
    /// the case it was given.
    Exponent {
        plus: bool,
        letter: char,
    },
    Fraction,
    Date(String),
    /// `[h]`, `[mm]`, `[ss]` — a unit that counts past its own wrap.
    Elapsed(String),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Comparison {
    Less,
    LessOrEqual,
    Greater,
    GreaterOrEqual,
    Equal,
    NotEqual,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Condition {
    pub operator: Comparison,
    pub value: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Kind {
    Number,
    Date,
    Text,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Section {
    pub tokens: Vec<Token>,
    /// `[Red]` and the rest, as written; the caller decides what red is.
    pub color: Option<String>,
    /// `[>=100]` — which numbers this section is for.
    pub condition: Option<Condition>,
    pub kind: Kind,
}

#[derive(Debug, Clone, PartialEq)]
pub struct NumberFormat {
    pub sections: Vec<Section>,
}

const COLORS: [&str; 8] = [
    "black", "blue", "cyan", "green", "magenta", "red", "white", "yellow",
];

/// Splits on the semicolons that separate sections, ignoring the quoted ones.
fn sections_of(code: &str) -> Vec<String> {
    let letters: Vec<char> = code.chars().collect();
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut quoted = false;
    let mut bracketed = false;
    let mut at = 0;

    while at < letters.len() {
        let letter = letters[at];

        if letter == '\\' && !quoted {
            current.push(letter);
            if let Some(next) = letters.get(at + 1) {
                current.push(*next);
            }
            at += 2;
            continue;
        }
        if letter == '"' {
            quoted = !quoted;
        }
        if !quoted && letter == '[' {
            bracketed = true;
        }
        if !quoted && letter == ']' {
            bracketed = false;
        }

        if letter == ';' && !quoted && !bracketed {
            parts.push(std::mem::take(&mut current));
            at += 1;
            continue;
        }

        current.push(letter);
        at += 1;
    }

    parts.push(current);
    parts
}

fn is_date_letter(letter: char) -> bool {
    matches!(letter.to_ascii_lowercase(), 'y' | 'm' | 'd' | 'h' | 's')
}

/// Whether the inside of a bracket is an elapsed unit rather than an
/// annotation: `[h]`, `[mm]`, `[ss]`.
fn is_elapsed(inside: &str) -> bool {
    !inside.is_empty()
        && inside
            .chars()
            .all(|letter| matches!(letter.to_ascii_lowercase(), 'h' | 'm' | 's'))
}

/// Whether a section formats a date.
///
/// Decided before anything else is read, because it is what makes `m` a month
/// and `/` a stroke. Letters inside quotes or brackets do not count: `0
/// "months"` is a number, however much it talks about time.
fn looks_like_date(body: &str) -> bool {
    let letters: Vec<char> = body.chars().collect();
    let mut bare = String::new();
    let mut at = 0;

    while at < letters.len() {
        match letters[at] {
            '"' => {
                at += 1;
                while at < letters.len() && letters[at] != '"' {
                    at += 1;
                }
                at += 1;
            }
            '\\' => at += 2,
            '[' => {
                let end = (at + 1..letters.len())
                    .find(|index| letters[*index] == ']')
                    .unwrap_or(letters.len());
                let inside: String = letters[at + 1..end.min(letters.len())].iter().collect();
                if is_elapsed(&inside) {
                    return true;
                }
                at = end + 1;
            }
            letter => {
                bare.push(letter);
                at += 1;
            }
        }
    }

    bare.chars().any(is_date_letter)
}

/// What a section says in brackets: its colour, its condition, its currency.
struct Bracketed {
    rest: String,
    color: Option<String>,
    condition: Option<Condition>,
    currency: Option<String>,
}

fn read_brackets(body: &str) -> Bracketed {
    let letters: Vec<char> = body.chars().collect();
    let mut rest = String::new();
    let mut found = Bracketed {
        rest: String::new(),
        color: None,
        condition: None,
        currency: None,
    };

    let mut at = 0;
    while at < letters.len() {
        // Quoted runs are stepped over: `"[" @ "]"` is a text format with
        // square brackets in it, not a colour and a condition.
        if letters[at] == '"' {
            rest.push('"');
            at += 1;
            while at < letters.len() && letters[at] != '"' {
                rest.push(letters[at]);
                at += 1;
            }
            if at < letters.len() {
                rest.push('"');
                at += 1;
            }
            continue;
        }

        if letters[at] == '\\' {
            rest.push('\\');
            if let Some(next) = letters.get(at + 1) {
                rest.push(*next);
            }
            at += 2;
            continue;
        }

        if letters[at] != '[' {
            rest.push(letters[at]);
            at += 1;
            continue;
        }

        let Some(end) = (at + 1..letters.len()).find(|index| letters[*index] == ']') else {
            rest.push('[');
            at += 1;
            continue;
        };

        let inside: String = letters[at + 1..end].iter().collect();
        let lower = inside.to_lowercase();

        // An elapsed unit is a token rather than an annotation, so it stays.
        if is_elapsed(&inside) {
            rest.push('[');
            rest.push_str(&inside);
            rest.push(']');
            at = end + 1;
            continue;
        }

        if COLORS.contains(&lower.as_str()) || is_numbered_color(&lower) {
            found.color = Some(inside);
            at = end + 1;
            continue;
        }

        if let Some(condition) = read_condition(&inside) {
            found.condition = Some(condition);
            at = end + 1;
            continue;
        }

        // `[$€-407]` states a currency and a locale; `[$-409]` states only
        // the locale, which decides month names in Excel and is ignored here.
        // The difference is whether there is a symbol before the dash.
        if let Some(money) = inside.strip_prefix('$') {
            let symbol = money.split('-').next().unwrap_or("");
            if !symbol.is_empty() {
                found.currency = Some(symbol.to_string());
            }
            at = end + 1;
            continue;
        }

        // Anything else in brackets — a locale on its own, a calendar — says
        // nothing this can act on and nothing worth showing.
        at = end + 1;
    }

    found.rest = rest;
    found
}

fn is_numbered_color(lower: &str) -> bool {
    match lower.strip_prefix("color") {
        Some(rest) => {
            let digits = rest.trim();
            !digits.is_empty() && digits.chars().all(|letter| letter.is_ascii_digit())
        }
        None => false,
    }
}

fn read_condition(inside: &str) -> Option<Condition> {
    // The two-letter operators first: `<=` read as `<` would leave an equals
    // sign where a number should be, and the condition would be thrown away.
    let (operator, rest) = if let Some(rest) = inside.strip_prefix(">=") {
        (Comparison::GreaterOrEqual, rest)
    } else if let Some(rest) = inside.strip_prefix("<=") {
        (Comparison::LessOrEqual, rest)
    } else if let Some(rest) = inside.strip_prefix("<>") {
        (Comparison::NotEqual, rest)
    } else if let Some(rest) = inside.strip_prefix('>') {
        (Comparison::Greater, rest)
    } else if let Some(rest) = inside.strip_prefix('<') {
        (Comparison::Less, rest)
    } else {
        (Comparison::Equal, inside.strip_prefix('=')?)
    };

    let value: f64 = rest.trim().parse().ok()?;
    Some(Condition { operator, value })
}

/// Runs of the same date letter are one token: `yyyy` is a year, not four.
fn date_run(letters: &[char], at: usize) -> String {
    let letter = letters[at].to_ascii_lowercase();
    let mut end = at;

    while letters
        .get(end + 1)
        .is_some_and(|next| next.to_ascii_lowercase() == letter)
    {
        end += 1;
    }

    letters[at..=end].iter().collect()
}

fn tokenise(body: &str, kind: Kind, currency: Option<String>) -> Vec<Token> {
    let letters: Vec<char> = body.chars().collect();
    let mut tokens: Vec<Token> = Vec::new();
    let mut at = 0;

    while at < letters.len() {
        let letter = letters[at];

        if letter == '"' {
            let end = (at + 1..letters.len())
                .find(|index| letters[*index] == '"')
                .unwrap_or(letters.len());
            tokens.push(Token::Literal(
                letters[at + 1..end.min(letters.len())].iter().collect(),
            ));
            at = end + 1;
            continue;
        }

        if letter == '\\' {
            tokens.push(Token::Literal(
                letters.get(at + 1).copied().unwrap_or(' ').to_string(),
            ));
            at += 2;
            continue;
        }

        if letter == '_' {
            tokens.push(Token::Pad(letters.get(at + 1).copied().unwrap_or(' ')));
            at += 2;
            continue;
        }

        if letter == '*' {
            tokens.push(Token::Fill(letters.get(at + 1).copied().unwrap_or(' ')));
            at += 2;
            continue;
        }

        if letter == '[' {
            let end = (at + 1..letters.len())
                .find(|index| letters[*index] == ']')
                .unwrap_or(letters.len());
            let inside: String = letters[at + 1..end.min(letters.len())].iter().collect();
            if is_elapsed(&inside) {
                tokens.push(Token::Elapsed(inside));
            }
            at = end + 1;
            continue;
        }

        if matches!(letter, '0' | '#' | '?') {
            tokens.push(Token::Digit(letter));
            at += 1;
            continue;
        }

        if letter == '.' {
            // A date's dot separates — `dd.mm.yy` — unless it is the point
            // before a fraction of a second, decided by what follows it.
            let fraction_of_second =
                kind == Kind::Date && letters.get(at + 1).is_some_and(|next| *next == '0');
            tokens.push(if kind == Kind::Date && !fraction_of_second {
                Token::Literal(".".to_string())
            } else {
                Token::Decimal
            });
            at += 1;
            continue;
        }

        if letter == '%' {
            tokens.push(Token::Percent);
            at += 1;
            continue;
        }

        if letter == '@' {
            tokens.push(Token::Text);
            at += 1;
            continue;
        }

        if matches!(letter, 'E' | 'e') && matches!(letters.get(at + 1), Some('+') | Some('-')) {
            tokens.push(Token::Exponent {
                plus: letters[at + 1] == '+',
                letter,
            });
            at += 2;
            continue;
        }

        if letter == ',' {
            // In a date a comma is punctuation — `mmm d, yyyy`. In a number
            // it is either grouping or scaling, and which one depends on what
            // follows, so the scaling ones are folded together afterwards.
            tokens.push(if kind == Kind::Date {
                Token::Literal(",".to_string())
            } else {
                Token::Group
            });
            at += 1;
            continue;
        }

        if letter == '/' && kind == Kind::Number {
            tokens.push(Token::Fraction);
            at += 1;
            continue;
        }

        if kind == Kind::Date {
            let ahead: String = letters[at..].iter().collect();
            let upper = ahead.to_uppercase();
            if upper.starts_with("AM/PM") {
                tokens.push(Token::Date(letters[at..at + 5].iter().collect()));
                at += 5;
                continue;
            }
            if upper.starts_with("A/P") {
                tokens.push(Token::Date(letters[at..at + 3].iter().collect()));
                at += 3;
                continue;
            }

            if is_date_letter(letter) {
                let run = date_run(&letters, at);
                at += run.chars().count();
                tokens.push(Token::Date(run));
                continue;
            }
        }

        tokens.push(Token::Literal(letter.to_string()));
        at += 1;
    }

    if let Some(symbol) = currency {
        tokens.insert(0, Token::Literal(symbol));
    }

    settle_commas(tokens)
}

/// What each comma in a number turns out to be.
///
/// Three different things wear the same character, and only their neighbours
/// tell them apart:
///
/// - **a separator**, between two places — `#,##0` shows thousands;
/// - **a divisor**, anywhere else among the places — `#,##0,` shows thousands
///   of thousands, and `#,.#,` divides twice on its way to the point;
/// - **a comma**, where no place has come yet — `,#` on a million and a bit
///   is `,1234567`, a literal comma and a number with no grouping at all.
///
/// The first of those is the only one anybody writes on purpose, which is why
/// the other two are easy to get wrong.
fn settle_commas(tokens: Vec<Token>) -> Vec<Token> {
    let mut settled: Vec<Token> = Vec::new();
    let mut seen_place = false;

    for (index, token) in tokens.iter().enumerate() {
        if *token != Token::Group {
            if matches!(token, Token::Digit(_)) {
                seen_place = true;
            }
            settled.push(token.clone());
            continue;
        }

        if !seen_place {
            settled.push(Token::Literal(",".to_string()));
            continue;
        }

        let before = index.checked_sub(1).and_then(|at| tokens.get(at));
        let after = tokens.get(index + 1);
        if matches!(before, Some(Token::Digit(_))) && matches!(after, Some(Token::Digit(_))) {
            settled.push(Token::Group);
            continue;
        }

        match settled.last_mut() {
            Some(Token::Scale(by)) => *by *= 1000.0,
            _ => settled.push(Token::Scale(1000.0)),
        }
    }

    settled
}

/// The AM/PM marker turns a 24-hour format into a 12-hour one.
///
/// Stated here rather than found later because it changes what `h` means, and
/// the hour is written long before the marker is reached.
pub fn is_twelve_hour(section: &Section) -> bool {
    section.tokens.iter().any(|token| match token {
        Token::Date(code) => {
            let upper = code.to_uppercase();
            upper == "AM/PM" || upper == "A/P"
        }
        _ => false,
    })
}

pub fn parse_format(code: &str) -> NumberFormat {
    let sections = sections_of(code)
        .into_iter()
        .map(|body| {
            let read = read_brackets(&body);
            let kind = if looks_like_date(&read.rest) {
                Kind::Date
            } else if read.rest.contains('@') {
                Kind::Text
            } else {
                Kind::Number
            };

            Section {
                tokens: tokenise(&read.rest, kind, read.currency),
                color: read.color,
                condition: read.condition,
                kind,
            }
        })
        .collect();

    NumberFormat { sections }
}
